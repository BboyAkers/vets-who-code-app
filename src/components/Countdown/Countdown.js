import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'

class Countdown extends Component {
  state = {
    days: null,
    hours: null,
    minutes: null,
    seconds: null,
    interval: null,
    time: null,
  }

  componentDidMount = () => {
    this.getTimeRemaining()
    const interval = setInterval(() => this.getTimeRemaining(), 1000)
    this.setState({ interval })
  }

  componentWillUnmount = () => {
    this.stopCountDown()
  }

  getTimeRemaining = () => {
    const { nextClass } = this.props
    const time = Date.parse(nextClass) - Date.parse(new Date())
    const seconds = Math.floor((time / 1000) % 60)
    const minutes = Math.floor((time / 1000 / 60) % 60)
    const hours = Math.floor((time / (1000 * 60 * 60)) % 24)
    const days = Math.floor(time / (1000 * 60 * 60 * 24))

    this.setState({
      days,
      hours,
      minutes,
      seconds,
      time,
    })
  }

  stopCountDown = () => {
    const { interval } = this.state
    clearInterval(interval)
  }

  render() {
    const { days, hours, minutes, seconds, time } = this.state
    return (
      <div>
        {time > 0 ? (
          <Fragment>
            <div className="container-countdown">
              <div className="countdown-box">
                <span className="counter">{days}</span>
                <h4>Days</h4>
              </div>
              <div className="countdown-box">
                <span className="counter">{hours}</span>
                <h4>Hours</h4>
              </div>
              <div className="countdown-box">
                <span className="counter">{minutes}</span>
                <h4>Minutes</h4>
              </div>
              <div className="countdown-box">
                <span className="counter">{seconds}</span>
                <h4>Seconds</h4>
              </div>
            </div>
          </Fragment>
        ) : (
          <div>
            <h3 className="countdown-message">Class Is In Session</h3>
          </div>
        )}
      </div>
    )
  }
}

Countdown.propTypes = {
  nextClass: PropTypes.string.isRequired, // String formatted 'March, 01 2021'
}

export default Countdown
