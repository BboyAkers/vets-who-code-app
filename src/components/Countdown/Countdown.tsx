import { useState, useEffect } from 'react'

interface CountdownProps {
  nextClass: string
}

type TLeadingZero = string | number

interface ITimeLeft {
  [key: string]: TLeadingZero
}

const addLeadingZero = (time: number): TLeadingZero => (time < 10 ? `0${time}` : time)

const calculateTimeLeft = (nextClass: string): ITimeLeft => {
  const difference: number = Date.parse(nextClass) - Date.parse(new Date().toString())
  let timeLeft: ITimeLeft = {}
  if (difference > 0) {
    timeLeft = {
      Days: Math.floor(difference / (1000 * 60 * 60 * 24)),
      Hours: addLeadingZero(Math.floor((difference / (1000 * 60 * 60)) % 24)),
      Minutes: addLeadingZero(Math.floor((difference / 1000 / 60) % 60)),
      Seconds: addLeadingZero(Math.floor((difference / 1000) % 60)),
    }
  }
  return timeLeft
}

const Countdown = ({ nextClass }: CountdownProps) => {
  const [timeLeft, setTimeLeft] = useState(calculateTimeLeft(nextClass))

  useEffect(() => {
    setTimeout(() => {
      setTimeLeft(calculateTimeLeft(nextClass))
    }, 1000)
  })

  const timerComponents = []

  Object.keys(timeLeft).forEach(interval => {
    if (!timeLeft[interval]) {
      return
    }
    timerComponents.push(
      <div className="countdown-box" key={interval}>
        <span className="counter">{timeLeft[interval]}</span>
        <h4>{interval}</h4>
      </div>
    )
  })

  return (
    <>
      <div className="container-countdown">
        {timerComponents.length ? (
          timerComponents
        ) : (
          <div>
            <h3 className="countdown-message">Class Is In Session</h3>
          </div>
        )}
      </div>
    </>
  )
}
export default Countdown
