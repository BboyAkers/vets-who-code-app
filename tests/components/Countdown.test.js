import React from 'react'
import Countdown from '../../src/components/Countdown'

describe('<Countdown />', () => {
  let wrapper

  beforeEach(() => wrapper = shallow(
    <Countdown />
  ))

  test('should render correctly', () => {
    wrapper.setState({
      days: 1,
      hours: 1,
      minutes: 27,
      seconds: 30
    })
    expect(wrapper).toMatchSnapshot()
  })
})
