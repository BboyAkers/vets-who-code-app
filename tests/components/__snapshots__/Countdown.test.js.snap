// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Countdown /> should render correctly 1`] = `
<div
  className="col-sm-6 event_counter_container text-center"
>
  <div
    className="container-countdown"
  >
    <div
      className="countdown-box"
    >
      <span
        className="counter"
      >
        1
      </span>
      <h4>
        Days
      </h4>
    </div>
    <div
      className="countdown-box"
    >
      <span
        className="counter"
      >
        1
      </span>
      <h4>
        Hours
      </h4>
    </div>
    <div
      className="countdown-box"
    >
      <span
        className="counter"
      >
        27
      </span>
      <h4>
        Minutes
      </h4>
    </div>
    <div
      className="countdown-box"
    >
      <span
        className="counter"
      >
        30
      </span>
      <h4>
        Seconds
      </h4>
    </div>
  </div>
  <GatsbyLink
    className="btn btn-charity-default"
    to="/apply"
  >
    Apply
  </GatsbyLink>
</div>
`;
